src/irmin-pack/inode.ml

```diff
@@ -487,6 +508,31 @@ struct
 
     let hash t = Lazy.force t.hash
 
+    let is_root t =
+      match t.v with
+      | Tree { depth; _ } -> depth = 0
+      | Values _ ->
+          (* When [t] is of tag [Values], then [t] is root iff [t] is stable. It
+             is implied by the following.
+
+             When [t] is stable, then [t] is a root, because:
+              - Only 2 functions produce stable inodes: [stabilize] and [empty].
+              - Only the roots are output of [stabilize].
```

Comment:
you are probably right, but I'm not 100% sure about this as `stabilize` is passed as argument to the recursive `add` function https://github.com/mirage/irmin/blob/master/src/irmin-pack/inode.ml#L564

Reply:
I've checked and `k` is only used to finalize the inode of the current call stack, it is never passed to recursive calls. The same holds of `remove`.

Thank you for checking that. I'm going to replace apply this diff everywhere applicable to make things clearer:
```diff
-  add ~depth:0 ~find ~copy ~replace:true t s v (stabilize ~find)
+  add ~depth:0 ~find ~copy ~replace:true t s v Fun.id |> stabilize ~find
```